README.md
# HW4

## Important Information
Aaron Yu
A16388367
[Netlify URL](https://guileless-kheer-b8f77b.netlify.app)

## Notes
The CRUD application uses the browser's localStorage to preserve changes. If the localStorage is empty (in other words, if this is the user's first time visiting the site with this localStorage session), three data entries are pre-loaded. All subsequent changes will be saved in localStorage. 

changelog.md
## Page Changes
Added JavaScript to highlight the name of the section that the user is currently scrolled to on the page in the navbar. Included optimizations like pre-computing all the section and name elements only on the first document load, searching the sections in reverse to minimize how often the classList of each nav element is changed, and only prompting a change when the active section has actually changed (user has scrolled out). 

Further, simplified the CSS, and changed how the page reflows when the screen size is smaller -- specifically, the sidebar nav becomes a top bar nav. On even smaller screens, the nav links are removed and replaced with a contact icon. This will eventually also be replaced with a hamburger menu instead, so the nav links can be preserved on smaller screens.

## Third-Party Script
Included **Google Analytics** in order to determine how users reach and engage with site so I can better optimize for conversion (getting people to contact me through LinkedIn or email).

Included **Lodash** so I could use the _.throttle() function to reduce the frequency at which my scroll handler for highlighting the name sections 

blog.js
import {showCRUDDialog} from './crud-dialog.js';
import {showConfirmDialog} from './customdialog.js';

const LOCAL_STORAGE_KEY = "blogposts";
let postCount = 0;

function createPost(_e, data) {
    let node = document.getElementById('blogposts');

    let noPostsMessage = node.querySelector('#blogposts > p');
    if(noPostsMessage) {
        noPostsMessage.remove();
    }
    
    let postContainer = document.createElement('section');
    let controlContainer = document.createElement('section');

    let title = document.createElement('h1');
    title.textContent = data.title;

    let date = document.createElement('h2');
    date.textContent = data.date;

    let summary = document.createElement('p');
    summary.textContent = data.summary;

    let editButton = document.createElement('button');
    editButton.textContent = 'Edit';
    editButton.addEventListener('click', function (e) {
        showCRUDDialog(
            "Edit Post", 
            (__e, data) => { if (data) editPost(e, data); }, 
            () => {
                return {
                    title: title.textContent,
                    date: date.textContent,
                    summary: summary.textContent
                };
        })
    });

    let deleteButton = document.createElement('button');
    deleteButton.textContent = 'Delete';
    deleteButton.addEventListener('click', function () {
        showConfirmDialog("Are you sure you want to delete?", (_e, confirmed) => {
            if(confirmed) removePost(postContainer);
        })
    });

    controlContainer.append(editButton, deleteButton);
    postContainer.append(title, date, summary, controlContainer);

    node.append(postContainer);
    postCount++;

    updateLocalStorage()
}

function editPost(e, data) {
    let editButton = e.target;
    let controlContainer = editButton.parentElement;
    let postContainer = controlContainer.parentElement;
    
    let title = postContainer.querySelector('h1');
    let date = postContainer.querySelector('h2');
    let summary = postContainer.querySelector('p');

    title.textContent = data.title;
    date.textContent = data.date;
    summary.textContent = data.summary;

    updateLocalStorage()
}

function removePost(post) {
    post.remove();
    postCount--;

    if(!postCount) {
        let node = document.getElementById('blogposts');
        let noPosts = document.createElement('p');
        noPosts.textContent = "No posts to show!";
        node.appendChild(noPosts);
    }

    updateLocalStorage();
}

function updateLocalStorage() {
    let blogposts = [];
    let node = document.getElementById('blogposts');
    let postContainers = node.querySelectorAll('#blogposts > section');

    postContainers.forEach((element) => {
        let title = element.querySelector('h1');
        let date = element.querySelector('h2');
        let summary = element.querySelector('p');

        blogposts.push({
            title: title.textContent,
            date: date.textContent,
            summary: summary.textContent
        });
    });

    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(blogposts));
    console.log(localStorage.getItem(LOCAL_STORAGE_KEY));

}

window.addEventListener('DOMContentLoaded', function () {
    let stored = JSON.parse(localStorage.getItem(LOCAL_STORAGE_KEY));

    if(!stored) {
        stored = [
            {title: "My First Blogpost", date: "2023-02-25", summary: "This is the text content of my first blogpost, which is about the creation of this CRUD application. Please enjoy reading!"}, 
            {title: "My Second Blogpost", date: "2023-02-26", summary: "This is the text content of my second blogpost, which is about the creation of this CRUD application. Please enjoy reading!"}, 
            {title: "My Third Blogpost", date: "2023-02-27", summary: "This is the text content of my third blogpost, which is about the creation of this CRUD application. Please enjoy reading!"}
        ];
    }
    stored.forEach((blogpost) => createPost(null, blogpost));

    document.getElementById('add-post').addEventListener('click', function () {
        showCRUDDialog("Create Post", function (e, data) {
            if (data) createPost(e, data);
        })
    });
})

crud-dialog.js
export function showCRUDDialog(title, rsltHandler, prefill=null) {
    const dialog = document.getElementById('crud-dialog');
    const dialogTitle = dialog.querySelector('h1');
    const form = dialog.querySelector('form');
    form.reset();

    if(prefill) {
        let prefillFields = prefill();
        form.querySelector('#title').value = prefillFields.title;
        form.querySelector('#date').value = prefillFields.date;
        form.querySelector('#summary').value = prefillFields.summary;
    }
    
    dialogTitle.textContent = title;
    dialog.showModal();

    dialog.addEventListener('close', (e) => {
        if(dialog.returnValue == "save") {
            let data = new FormData(form);
            rsltHandler(e, {
                title: data.get('title'), 
                date: data.get('date'), 
                summary: data.get('summary')
            });
        } else {
            rsltHandler(e, null);
        }
        
    }, {once: true});
}

customdialog.js
export function showAlertDialog (message) {
    const dialogMessage = document.querySelector('#alert-dialog > p');
    dialogMessage.textContent = message;

    const alertDialog = document.getElementById('alert-dialog');
    alertDialog.showModal();
}

export function showConfirmDialog (message, rsltHandler) {
    const dialogMessage = document.querySelector('#confirm-dialog > p');
    dialogMessage.textContent = message;

    const confirmDialog = document.getElementById('confirm-dialog');
    confirmDialog.showModal();
    confirmDialog.addEventListener('close', (e) => rsltHandler(e, confirmDialog.returnValue === "true"), {once: true});
}

export function showPromptDialog (prompt, rsltHandler) {
    const dialogMessage = document.querySelector('#prompt-dialog > p');
    dialogMessage.textContent = prompt;

    const promptDialog = document.getElementById('prompt-dialog');
    promptDialog.showModal();

    const form = document.querySelector('#prompt-dialog > form');
    promptDialog.addEventListener('close', (e) => {
        let response = new FormData(form).get('response');
        if(promptDialog.returnValue == 'OK') {
            rsltHandler(e, response);
        } else {
            rsltHandler(e, '');
        }
    }, {once: true});
}

export function sanitize_xss(strings, str) {
    return strings[0] + DOMPurify.sanitize(str);
}




scripts/main.js
const NAV_HIGHLIGHT_UPDATE_PERIOD = 150;
window.addEventListener('DOMContentLoaded', () => {
    // Get sections and nav links
    const sections = [...document.querySelectorAll('main > section')].reverse(); // Search sections from end
    const navLinks = document.querySelectorAll('.nav-main a');

    // Map section IDs to the corresponding nav link, to avoid repeated DOM queries
    let idToNav = {};
    sections.forEach((section) => {
        idToNav[section.id] = document.querySelector(`.nav-main a[href*=${section.id}]`);
    })

    highlightActiveNav();

    function highlightActiveNav () {
        // Media breakpoint for switching to topbar (needs additional offset on scroll height)
        let offsetHeight = window.innerWidth <= 984 ? 4 * parseFloat(getComputedStyle(document.documentElement).fontSize) : 0; 
        sections.every((section) => {
            if(window.scrollY + offsetHeight >= section.offsetTop // Scrolled past the top of the section
                || window.scrollY + window.innerHeight + 1 >= document.body.scrollHeight) { // Or past last section
                // Only change classList if we need to (scrolled to another section)
                if(!idToNav[section.id].classList.contains('active')){
                    navLinks.forEach((navlink) => navlink.classList.remove('active'));
                    idToNav[section.id].classList.add('active');
                }
                return false;
            }
            return true;
        });
    }

    window.addEventListener('scroll', _.throttle(highlightActiveNav, NAV_HIGHLIGHT_UPDATE_PERIOD));
})

styledblog.js
import {showCRUDDialog} from './crud-dialog.js';
import {showConfirmDialog} from './customdialog.js';

const LOCAL_STORAGE_KEY = "blogposts";
let postCount = 0;

function createPost(_e, data) {
    let node = document.getElementById('blogposts');

    let noPostsMessage = node.querySelector('#blogposts > p');
    if(noPostsMessage) {
        noPostsMessage.remove();
    }
    
    let postContainer = document.createElement('section');
    let controlContainer = document.createElement('section');

    let title = document.createElement('h1');
    title.textContent = data.title;

    let date = document.createElement('h2');
    date.textContent = data.date;

    let summary = document.createElement('p');
    summary.textContent = data.summary;

    let editButton = document.createElement('button');
    const editIcon = `<svg class="icon"><use xlink:href="/media/icons/solid.svg#pencil"></use></svg>`;
    editButton.innerHTML = editIcon;
    editButton.addEventListener('click', function (e) {
        showCRUDDialog(
            "Edit Post", 
            (__e, data) => { if (data) editPost(e, data); }, 
            () => {
                return {
                    title: title.textContent,
                    date: date.textContent,
                    summary: summary.textContent
                };
        })
    });

    let deleteButton = document.createElement('button');
    const deleteIcon = `<svg class="icon"><use xlink:href="/media/icons/solid.svg#trash-can"></use></svg>`;
    deleteButton.innerHTML = deleteIcon;
    deleteButton.addEventListener('click', function () {
        showConfirmDialog("Are you sure you want to delete?", (_e, confirmed) => {
            if(confirmed) removePost(postContainer);
        })
    });

    controlContainer.append(editButton, deleteButton);
    postContainer.append(title, date, summary, controlContainer);

    node.append(postContainer);
    postCount++;

    updateLocalStorage()
}

function editPost(e, data) {
    let editButton = e.target;
    let controlContainer = editButton.parentElement;
    let postContainer = controlContainer.parentElement;
    
    let title = postContainer.querySelector('h1');
    let date = postContainer.querySelector('h2');
    let summary = postContainer.querySelector('p');

    title.textContent = data.title;
    date.textContent = data.date;
    summary.textContent = data.summary;

    updateLocalStorage()
}

function removePost(post) {
    post.remove();
    postCount--;

    if(!postCount) {
        let node = document.getElementById('blogposts');
        let noPosts = document.createElement('p');
        noPosts.textContent = "No posts to show!";
        node.appendChild(noPosts);
    }

    updateLocalStorage();
}

function updateLocalStorage() {
    let blogposts = [];
    let node = document.getElementById('blogposts');
    let postContainers = node.querySelectorAll('#blogposts > section');

    postContainers.forEach((element) => {
        let title = element.querySelector('h1');
        let date = element.querySelector('h2');
        let summary = element.querySelector('p');

        blogposts.push({
            title: title.textContent,
            date: date.textContent,
            summary: summary.textContent
        });
    });

    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(blogposts));
    console.log(localStorage.getItem(LOCAL_STORAGE_KEY));

}

window.addEventListener('DOMContentLoaded', function () {
    let stored = JSON.parse(localStorage.getItem(LOCAL_STORAGE_KEY));

    if(!stored) {
        stored = [
            {title: "My First Blogpost", date: "2023-02-25", summary: "This is the text content of my first blogpost, which is about the creation of this CRUD application. Please enjoy reading!"}, 
            {title: "My Second Blogpost", date: "2023-02-26", summary: "This is the text content of my second blogpost, which is about the creation of this CRUD application. Please enjoy reading!"}, 
            {title: "My Third Blogpost", date: "2023-02-27", summary: "This is the text content of my third blogpost, which is about the creation of this CRUD application. Please enjoy reading!"}
        ];
    }
    stored.forEach((blogpost) => createPost(null, blogpost));

    document.getElementById('add-post').addEventListener('click', function () {
        showCRUDDialog("Create Post", function (e, data) {
            if (data) createPost(e, data);
        })
    });
})